File: src/render/fit.ts
// ============================================================
// ClaudeHub - 状态栏宽度自适应
// ============================================================
//
// 状态栏的任何一行如果超过终端列宽，终端会把它折成两行，
// 而 Claude Code 按“逻辑行”做光标记账，于是后续重绘就会错位错乱
// （Claude Code 2.1.141 修过一版：多行状态栏任一行超宽会丢行/错乱）。
// 这里从 Claude Code 注入的环境变量 COLUMNS 拿到终端列宽，自己把每行
// 裁到列宽以内，从源头消掉折行。

import { RESET } from './colors.js';

/** ANSI 转义序列（本项目只输出 CSI SGR 形式） */
const ANSI_SGR_RE = /\x1b\[[0-9;]*m/g;

/** 全宽（双列）字符范围：假名、韩文、CJK、全角形 */
const FULLWIDTH_RE =
  /[ᄀ-ᅟ⺀-⻿　-ヿ㐀-䶿一-鿿가-힯豈-﫿︰-﹯＀-｠]/;

const ELLIPSIS = '…';

/** 单个字符（码点）的显示宽度 */
function charWidth(ch: string): number {
  // 星形平面字符（emoji 等 surrogates）按 2 列计
  if (ch.length > 1) return 2;
  return FULLWIDTH_RE.test(ch) ? 2 : 1;
}

/** 字符串的可见列宽（忽略 ANSI 转义序列） */
export function visibleWidth(text: string): number {
  if (!text) return 0;
  const plain = text.replace(ANSI_SGR_RE, '');
  let width = 0;
  for (const ch of [...plain]) width += charWidth(ch);
  return width;
}

/** 把一行裁到指定列宽（保留转义序列；末尾补省略号并归位颜色） */
export function clipLine(line: string, columns: number): string {
  if (columns <= 0) return line;
  if (visibleWidth(line) <= columns) return line;

  const chars = [...line];
  let out = '';
  let width = 0;

  for (let i = 0; i < chars.length; i++) {
    const ch = chars[i];

    // 转义序列整段复制，不占列宽
    if (ch === '\x1b') {
      let seq = ch;
      for (let j = i + 1; j < chars.length; j++) {
        seq += chars[j];
        if (chars[j] === 'm') { i = j; break; }
      }
      out += seq;
      continue;
    }

    const w = charWidth(ch);
    // 留 1 列给省略号
    if (width + w > columns - 1) return `${out}${ELLIPSIS}${RESET}`;
    out += ch;
    width += w;
  }

  return out;
}

/** 读取 Claude Code 注入的终端列宽；拿不到返回 0（不裁剪） */
export function getTerminalColums(): number {
  const raw = process.env?.COLUMNS;
  if (!raw) return 0;
  const num = Number(raw);
  if (Number.isNaN(num) || num <= 0) return 0;
  return Math.floor(num);
}

/** 逐行裁剪，并按 maxLines 限行（maxLines <= 0 表示不限） */
export function fitLines(lines: string[], columns: number, maxLines: number = 0): string[] {
  let out = lines.map((line: string) => clipLine(line, columns));
  if (maxLines > 0 && out.length > maxLines) out = out.slice(0, maxLines);
  return out;
}

File: hud-rs/src/fit.rs
// ============================================================
// 列宽自适应
// ============================================================
//
// 状态栏任何一行超过终端列宽，终端会折行，而 Claude Code 按逻辑行记账，
// 后续重绘就会错位错乱。这里按 COLUMNS 把每行裁到列宽以内。

pub const RESET: &str = "\x1b[0m";
pub const DIM: &str = "\x1b[2m";
pub const RED: &str = "\x1b[31m";
pub const GREEN: &str = "\x1b[32m";
pub const YELLOW: &str = "\x1b[33m";
pub const MAGENTA: &str = "\x1b[35m";
pub const CYAN: &str = "\x1b[36m";
pub const BRIGHT_BLUE: &str = "\x1b[94m";
pub const BRIGHT_MAGENTA: &str = "\x1b[95m";

const ELLIPSIS: char = '…';

/// 判断是否为全宽（双列）字符
fn is_wide(c: char) -> bool {
    matches!(c as u32,
        0x1100..=0x115F |     // 谚文字母
        0x2E80..=0x303E |     // CJK 部首、假名标点
        0x3041..=0x33FF |     // 假名、CJK 兼容
        0x3400..=0x4DBF |     // CJK 扩展 A
        0x4E00..=0x9FFF |     // CJK 统一表意
        0xA000..=0xA4CF |     // 彝文
        0xAC00..=0xD7A3 |     // 谚文音节
        0xF900..=0xFAFF |     // CJK 兼容表意
        0xFE30..=0xFE6F |     // CJK 兼容形式
        0xFF00..=0xFF60 |     // 全角形式
        0xFFE0..=0xFFE6 |
        0x1F300..=0x1F64F |   // 表情符号
        0x1F900..=0x1F9FF |
        0x20000..=0x3FFFD
    )
}

/// 字符串的可见列宽（忽略 ANSI 转义序列，CJK / emoji 按 2 列）
pub fn visible_width(s: &str) -> usize {
    let mut width = 0usize;
    let mut chars = s.chars().peekable();
    while let Some(c) = chars.next() {
        if c == '\x1b' {
            // 跳过 CSI 序列：ESC [ ... 终止字符
            if chars.peek() == Some(&'[') {
                chars.next();
                for c2 in chars.by_ref() {
                    if c2.is_ascii_alphabetic() {
                        break;
                    }
                }
            }
            continue;
        }
        // 控制字符不占宽度
        if (c as u32) < 0x20 || (0x7f..=0x9f).contains(&(c as u32)) {
            continue;
        }
        width += if is_wide(c) { 2 } else { 1 };
    }
    width
}

/// 把一行裁到指定列宽（保留转义序列，超宽处补省略号并归位颜色）
pub fn clip_line(line: &str, columns: usize) -> String {
    if columns == 0 || visible_width(line) <= columns {
        return line.to_string();
    }

    let mut out = String::new();
    let mut width = 0usize;
    let mut chars = line.chars().peekable();

    while let Some(c) = chars.next() {
        // 转义序列整段复制，不占列宽
        if c == '\x1b' {
            out.push(c);
            if chars.peek() == Some(&'[') {
                out.push(chars.next().unwrap());
                for c2 in chars.by_ref() {
                    out.push(c2);
                    if c2.is_ascii_alphabetic() {
                        break;
                    }
                }
            }
            continue;
        }

        let w = if (c as u32) < 0x20 || (0x7f..=0x9f).contains(&(c as u32)) {
            0
        } else if is_wide(c) {
            2
        } else {
            1
        };

        // 留 1 列给省略号
        if width + w > columns.saturating_sub(1) {
            out.push(ELLIPSIS);
            out.push_str(RESET);
            return out;
        }
        out.push(c);
        width += w;
    }

    out
}

/// 读取 Claude Code 注入的终端列宽；拿不到返回 0（不裁剪）
pub fn terminal_columns() -> usize {
    std::env::var("COLUMNS")
        .ok()
        .and_then(|v| v.trim().parse::<usize>().ok())
        .filter(|n| *n > 0)
        .unwrap_or(0)
}
